IDE — Enable keyboard navigation, move alternativeText to individual files and chat

joshfarrant opened the pull request:
## Summary

Enables keyboard navigation of the tabs within the IDE component. To enable this, it was necessary to remove the top-level `alternativeText` from the IDE component in favour of `alternativeText` per file, and for the chat.

## List of notable changes:

- Removed `alternativeText` prop from `IDE`
- Added `alternativeText` prop to `IDE.Chat`
- Added `alternativeText` property to `IDEEditorFile` type
- Added `useTabs` hook to manage tab behaviour according to [the Tabs Pattern ARIA APG](https://www.w3.org/WAI/ARIA/apg/patterns/tabs/). Everything listed in that document has been implemented, except:
  - <kbd>Shift</kbd> + <kbd>F10</kbd>: Open associated popup menu (not relevant)
  - <kbd>Delete</kbd>: Tab deletion (not relevant)
- Refactored `IDE.Editor` line animation to simplify logic
- Updated `IDE` documentation with `alternativeText`
- Added _Accessibility_  section to `IDE` documentation

## What should reviewers focus on?

- Check that the keyboard navigation works as expected
- Check that animations run correctly, can be replayed, etc

## Steps to test:

- [Link to docs](https://primer-0b680965b8-26139705.drafts.github.io/brand/components/IDE)
- [Link to Storybook](https://primer-0b680965b8-26139705.drafts.github.io/brand/storybook/?path=/story/components-ide--default)

## Supporting resources (related issues, external links, etc):

- Closes https://github.com/github/primer/issues/3766
- Closes https://github.com/github/primer/issues/3768
- Closes https://github.com/github/primer/issues/3769
- Closes https://github.com/github/primer/issues/3770

## Contributor checklist:

- [x] All new and existing CI checks pass
- [x] Tests prove that the feature works and covers both happy and unhappy paths
- [x] Any drop in coverage, breaking changes or regressions have been documented above
- [x] UI Changes contain new visual snapshots (generated by adding `update snapshots` label to the PR)
- [x] All developer debugging and non-functional logging has been removed
- [x] Related issues have been referenced in the PR description

## Reviewer checklist:

- [ ] Check that pull request and proposed changes adhere to our [contribution guidelines](../../CONTRIBUTING.md) and [code of conduct](../../CODE_OF_CONDUCT.md)
- [ ] Check that tests prove the feature works and covers both happy and unhappy paths
- [ ] Check that there aren't other open Pull Requests for the same update/change

## Demo:

https://github.com/user-attachments/assets/08f18b0c-906d-4976-abfb-1efc8644fcea

rezrah commented on packages/react/src/hooks/useTabs.ts:
```diff
@@ -0,0 +1,229 @@
+    tabs: new Set(),
+  })
+
+  const setState = useCallback(
```
Could we use something besides `setState` for this? That feels more a defacto reserved term and is needing you to prepend the usual setState with `_`. 

Perhaps call this `setTabState` instead?

joshfarrant replied:
Happy to change it, but the choice of `setState` was intentional here as I want to encourage use of my custom `setState` wrapper and discourage use of the original `_setState` since using the raw `_setState` will prevent `onTabActivate` from being called

joshfarrant replied:
Just checking that I've not misunderstood the comment. I'd be happy to change to the following if this is what you were suggesting:

```tsx
  const [tabState, _setTabState] = useState<TabState>({
    activeTab: defaultTab || null,
    focusedTab: null,
    tabs: new Set(),
  })

  const setTabState = useCallback(
```

rezrah replied:
Yeah that works, thanks 👍

joshfarrant replied:
Cool, updated c99b1d4a35daa8fa3448053fc0531a9424549899

rezrah commented on packages/react/src/hooks/useTabs.ts:
```diff
@@ -0,0 +1,229 @@
+    [setState],
+  )
+
+  const getNextTabIdx = useCallback((idx: number, tabs: string[]) => (idx + 1) % tabs.length, [])
```
Nit.. could we avoid abbreviation on the variable name please? Fine to keep it for argument name if you want though. Just for readability.

```suggestion
  const getNextTabIndex = useCallback((idx: number, tabs: string[]) => (idx + 1) % tabs.length, [])
```

joshfarrant replied:
No prob 👍

bb3ef8d10174b4ace33faa1944c4a2283cf19aa0

rezrah commented on .changeset/warm-ads-return.md:
````diff
@@ -0,0 +1,37 @@
+
+⚠️ Breaking changes
+
+The `alternativeText` prop on the `IDE` component has been removed in favour of more granular descriptive text.
+
+Alternative text should now be provided in the following ways:
+
+- Via the `alternativeText` prop on the `IDE.Chat` component.
+  ```tsx
+  <IDE>
+    <IDE.Chat
+      alternativeText="A user asks how to concatenate arrays in JavaScript, Copilot demonstrates using the concat method, and the user confirms it worked."
+      // ...
+    />
+  </IDE>
+  ```
+- Via the `alternativeText` property on each file provided to the `IDE.Editor`.
+  ```tsx
+  <IDE>
+    <IDE.Editor
+      files={[
+        {
+          name: 'index.js',
+          alternativeText: 'TypeScript sentiment analysis function with D3.js visualization.',
+          // ...
+        },
+      ]}
+    />
+  </IDE>
+  ```
````
Nice. Could we switch the latter part for before/afters to make this easier to parse?

E.g.

The `alternativeText` prop on the `IDE` component has been removed in favour of more granular descriptive text.


<table width="100%">
<tr>
<th> Before</th>
</tr>
<tr>
<td valign="top">

```tsx
<IDE alternativeText="A user asks how to concatenate arrays in JavaScript, Copilot demonstrates using the concat method, and the user confirms it worked.">
  <IDE.Chat />
</IDE>
```

 </td>
</tr>
<tr>
<th> After</th>
</tr>
<tr>
<td valign="top">

```tsx
<IDE>
  <IDE.Chat alternativeText="A user asks how to concatenate arrays in JavaScript, Copilot demonstrates using the concat method, and the user confirms it worked." />
</IDE>
```
</td>
</tr>
</table>


<table width="100%">
<tr>
<th> Before</th>
</tr>
<tr>
<td valign="top">

```tsx
<IDE alternativeText="TypeScript sentiment analysis function with D3.js visualization.">
  <IDE.Editor
    files={[
      {
        name: 'index.js'
      },
    ]}
  />
</IDE>
```

 </td>
</tr>
<tr>
<th> After</th>
</tr>
<tr>
<td valign="top">

```tsx
<IDE>
  <IDE.Editor
    files={[
      {
        name: 'index.js',
        alternativeText: 'TypeScript sentiment analysis function with D3.js visualization.',
        // ...
      },
    ]}
  />
</IDE>
```

</td>
</tr>
</table>


Can't add as a suggestion, but you can inspect the comment if you need the Markdown for it.

joshfarrant replied:
Nice, updated d83288974d9a812bae337849520acf2814232659

rezrah commented on packages/react/src/IDE/IDE.tsx:
```diff
@@ -353,234 +337,188 @@ const _Editor = memo(
-          <div className={styles['IDE__Editor-tabs']} ref={tabsRef} data-testid={testIds.editorTabs}>
-            {files.map((file, index) => {
+          <div className={styles['IDE__Editor-tabs']} data-testid={testIds.editorTabs} {...tabs.getTabListProps()}>
+            {files.map((file, i) => {
```
Nit and just curious, was there a reason for removing the verbose name for index here?

joshfarrant replied:
No there wasn't, and I hadn't realised I'd done it to be honest. I refactored around there a couple of times and that must have just slipped through. If you'd prefer `index` I can update it?

joshfarrant replied:
Updated anyway 🙂

f6b1b2a44d81ab9499d07bd0480c34319f238e68

rezrah reviewed (approved):
Looks great, thanks @joshfarrant. I haven't done a super thorough review of the changes to IDE.tsx as the diff was a bit much 🫠... The examples all seem to work well though, nice one.

joshfarrant commented on packages/react/src/IDE/IDE.tsx:
```diff
@@ -353,234 +337,188 @@ const _Editor = memo(
                     <img
                       className={styles['IDE__Editor-tab-icon']}
-                      alt={`Logo for ${language}`}
+                      alt=""
```
I spoke to Patrick about this as the previous behaviour felt a bit janky when announced by a screen reader.

NVDA, for example, would announce "logo for ts sentiments.ts". We agreed that in this case the icon was purely decorative and didn't need to be announced as the "logo for ts" was fairly redundant when read alongside the filename.

joshfarrant commented:
Thanks for the review @rezrah 💖

The large diff to IDE.tsx couldn't be avoided unfortunately. I'd highly recommend viewing the changes with _Hide whitespace_ enabled as that will make it a bit easier to parse, since some of the larger diffs were due to wrappers being added around pre-existing HTML.

Other than that, the largest change was a refactor of the animation triggering logic. The core approach has stayed the same, I just added the ability to re-trigger it programatically and removed a bit of duplication around managing the delay of triggered elements.

If you've got 10 mins I'd be more than happy to walk you through it 🙂
